PX-RunC DaemonSet merge

zoxpx opened the pull request:
Summary of changes:
* introduced new `px-runcds` (PX RunC DaemonSet) binary
* updated Makefile (static build, apline-based containers)
* updated WebSvc (using Gorilla-scheme for form processing, RunC modifications, new usage page, PWX-3815 fix)
* introduced vendoring (extl. sources pulled during build)

harsh-px commented on px-mon/px-mon.go:
```diff
@@ -11,13 +11,13 @@ import (
 	dockerclient "github.com/fsouza/go-dockerclient"
```
At this point, no product is using px-mon.go. So it's dead code. Just remove it from here and build.

zoxpx replied:
Oh.. so both `px-init` and `px-mon` are dead-code?
Ok, will remove form the build.

harsh-px commented on px-oci-mon/main.go:
```diff
@@ -0,0 +1,259 @@
+func installPxFromOciImage(di *utils.DockerInstaller, imageName string, cfg *utils.SimpleContainerConfig) error {
+	logrus.Info("Downloading Portworx...")
+
+	err := di.PullImage(imageName)
```
Can be changed to `if err := di.PullImage(imageName); err != nil`

Same in other parts of the file.

harsh-px commented on px-oci-mon/main.go:
```diff
@@ -0,0 +1,259 @@
+
+	di, err := utils.NewDockerInstaller(os.Getenv("REGISTRY_USER"), os.Getenv("REGISTRY_PASS"))
+	if err != nil {
+		logrus.WithError(err).Error("Could not 'talk' to Docker")
```
nit: Why is the 'talk' in quotes?

harsh-px commented on px-spec-websvc/px-spec-websvc.go:
```diff
@@ -3,69 +3,86 @@ package main
-	Openshift  string
-	PxImage    string
-	MasterLess bool
+	Cluster     string `schema:"c"    deprecated:"cluster"`
```
Why deprecate? Based on the end user, he/she might prefer using short or long options.

zoxpx replied:
Let's please not have 2 standards -- stick with one, and deprecate the other.
The "long" parameters are IMO not really helping -- they make the URLs really long, and they are not consistent with our "normal" PX-parameters.

harsh-px commented on px-spec-websvc/templates/k8s-master-worker-response.gtpl:
```diff
@@ -122,15 +128,25 @@ spec:
+        {{- if .IsRunC}}
+        - name: proc1nsmount
+          hostPath:
+            path: /proc/1/ns
```
Most companies won't allow mount this namespace.

harsh-px commented on px-oci-mon/main.go:
```diff
@@ -0,0 +1,259 @@
+		os.Exit(-1)
+	}
+
+	logrus.Info("Stopping Portworx service (if any)")
```
One of the use cases of moving to runc is not get affected when k8s restarts the portworx pod container.

Because you are stopping and restarting portworx unconditionally here, the above use case is violated.

Here is an actual issue we saw.

1. k8s minion/node goes into disk pressure at 75% disk usage. 
2. At this point, kubelet goes into a node eviction phase where it picks ALL pods running on the node and starts deleting them. 
3. So the portworx pod gets deleted. 
4. Since portworx runs as a daemonset, a replacement pod shows up right away

In the runc world, steps 3 and 4 will be the `px-oci-mon/main.go` (Correct me if I'm wrong). So every time, the disk pressure goes 75% on a node, PX will flap.

Actually, in general, every docker restart will flap PX runc. Am I missing something?

zoxpx replied:
Yes -- I know, and I pointed it out a few times this code will need to be improved, and "restarts" be made more "intelligent" (conditional).

However, I'd prefer to merge the code first (as there is already a lot of code pending), and then fix this (actually, improve the code) on `master` -branch

zoxpx replied:
NOTE -- filed https://portworx.atlassian.net/browse/PWX-3927 to address this issue.

harsh-px commented on px-oci-mon/main.go:
```diff
@@ -0,0 +1,259 @@
+	err = runExternal("/bin/sh", "-c",
+		`systemctl disable portworx`)
+	if err != nil {
+		logrus.WithError(err).Warn("Could not disable Portworx service (continuing)")
```
Looks like the approach here is best-effort.

if we failed to disable the service, let's not move forward. A service that isn't disabled will attempt to spin up on the next systemd start. If we ignore the error here, it will give the end user a false positive that portworx is uninstalled from the system, while in reality, the service is still enabled.

zoxpx replied:
ok.

zoxpx replied:
Actually -- the service files, and the binaries will be removed, so the service won't be able to start, but let's not risk weird service errors -- will change this to critical failure.

harsh-px commented on px-init/px-init.go:
```diff
@@ -13,7 +13,7 @@ func enableSharedMounts() error {
 
```
Just like px-mon.go, remove px-init from source and build. dead code.

harsh-px commented on px-oci-mon/main.go:
```diff
@@ -0,0 +1,259 @@
+		usage("Command " + os.Args[1] + " not supported")
+	}
+
+	// CHECKME: Should we always go to sleep, or can we also exit
```
It appears this is a entrypoint of the k8s pod container. So we cannot exit. So remove the comment.

harsh-px commented on px-oci-mon/utils/installer.go:
```diff
@@ -0,0 +1,188 @@
+		RemoveVolumes: true,
+		Force:         true,
+	})
+	logrus.WithError(err).Debug("Old container removed")
```
The logrus string should be `failed to remove old container`.

harsh-px commented on px-oci-mon/utils/installer.go:
```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return nil, fmt.Errorf("Error inspecting container '%s': %s", id, err)
+	}
+	logrus.Debugf("COnFIG:%+v", cconf)
```
Typo: `COnFIG`

zoxpx commented on Makefile:
```diff
@@ -1,30 +1,115 @@
-	go build -o px-spec-websvc/px-spec-websvc px-spec-websvc/px-spec-websvc.go
+# VARIABLES
+#
+ifndef DOCKER_HUB_REPO
```
Sounds good -- previously we were using different (unique) variable, so it wasn't a big deal -- but now the risk is higher.

zoxpx commented:
Addressed most of the issues with follow-up checkins.

Unaddressed issues are as follows:
* docker restart still reinstalls/restarts the PX-RunC (filed separate [PWX-3927](https://portworx.atlassian.net/browse/PWX-3927) to address this issue)
* WebSVC - still deprecating the "long" parameters
* still mounting the `/proc/1/ns` (required to install/start the external PX-RunC service)

Checked verbally w/ Harsh, and he was OK to proceed to code-merge.

harsh-px commented:
Discussed offline with @zoxpx Changes look good now. We have filed a ticket for fixing the dependency on docker bouncing.
